packages/xstate-react/src/types.ts

```diff
@@ -46,3 +46,8 @@ export interface ActorRefLike<TEvent extends EventObject, TEmitted = any>
 }
 
 export type MaybeLazy<T> = T | (() => T);
+
+export interface PayloadSender<TEvent extends EventObject> {
+  (event: TEvent | TEvent['type']): void;
+  (event: TEvent['type'], payload: Omit<TEvent, 'type'>): void;
```

Comment:
This is not correct for a couple of reasons.

It now kinda accepts `any` on the second position because `Omit` is just a wrapper around a combination `Pick`, `Exclude`, and `keyof`. Problem is that `keyof` used on the union (and `TEvent` is usually one) only returns common keys and usually `type` is the only known key - so in the end what we get here is `Pick<TEvent, never>` (when this gets computed based on some passed-in concrete generic parameter). See this [TS playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKIDcLgPICMBWECYyA3gFDLJgCeADhAFzIDOYUoA5gNzkC+5cqEixEKAApwaAGwD2cACYBlLAugAeACoYsJCAA9IIBSzSYcBImAB8ZSsgAUEc2GbaXyAD7J3ugNoA5LQMAQC6AJTM6LLACrxUTi5uOuCBwRBhADTIdFJyiszYALbAYFopYNlB9BnWkcjRsbwC5OlmuqYAvHZUVOnMAQBi2NgB9lRwICwA7tDMIACuRbjQ-F49vf3IAQBCAIIASmO9yLJgABZzrOxc-IJqCNJwUCgIslMkLKrMkjLyyqoNBUWNZBF9jA5docAtlSMhcJNESwBpA2AFkHxwoI2gBpCA0WQwbAgACqIGA72Q3QA1vjCe1wCxBEA)

To overcome this problem we could tweak this to use a distributive version of the `Omit`, which is this:
```ts
type DistributiveOmit<T, K extends keyof T> = T extends any ? Omit<T, K> : never
```
A version with this would at least accept payloads limited to a particular set but it would allow incorrect combinations because there is no relation specified between both arguments. See this [TS playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKIDcLgPICMBWECYyA3gFDLJgCeADhAFzIDOYUoA5gNzkC+5crQbIAIsDYdcAVzDBM2ALbAwAHgAqAGmQBpZBAAekEABMWyANYQaAexjJ1APmQBeB-qNYzyOCBrIAfmQlFQ1tHWdmEAhMKEFQSFhEFAAFOBoAGxs4EwBlL2gNDCwSQ2NvYpwCIjBnCioAChiS5nVKkgAfB3aAbQByYQg+gF0ASmZ0G2ATXkbm8Fbegfoh4e06dKyc5nFJYBk5BWU1NsxwbWWGPsdx5Enp3gEhFbQzsHM3eqpqFeY+gDFsNg+pRvr4WAB3aBRaSKXDQfjILpfKiDP4AIQAggAlEHfZA2MAAC2hrHYXH4ghMRAycCgKAQNhAbFYXmYaUy2TyBSgqnaLEcggA9EKCTYbHQWNpQIyoPTiOQWF4Gn0sbjtKQfMyoVBmAAWABMyD4o0E5CAA)

To properly type this we'd have to introduce a generic for the second overload:
```ts
<K extends TEvent["type"]>(event: K, payload: Omit<Extract<TEvent, { type: K }>, 'type'>): void;
```
See this [TS playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKIDcLgPICMBWECYyA3gFDLJgCeADhAFzIDOYUoA5gNzkC+5cqEixEKAApwaAGwD2cACYBlLAugAeACoYsJCAA9IIBSzSYcBImAB8ZSsgAUEc2GbaXyAD7J3ugNoA5LQMAQC6AJTM6LLACrxU6gDSyAZGJj464H4ARMEQ2aHWTi7MiQA0yHRScorM2AC2wGDqqIZQiM2+4BWk1PRMyMl81hVB-QHWkcjRsbwC5HlmuqYAvHZUVHnMAQBi2NgB9lRwICwA7tDMIACu9bjQ-F7rG1vIAQBCAIIASocbyLIwAALS6sdhcfiCNQIaRwKAoBCyU4kFiqZiSGTyZSqDSZMAsayCVHGBwfH4BHoA4GggKKALIPjhcjEhSkr6-Sknc6ggAsACYGUyWWzyZTXmSOVSQVBtnTBeQgA).

Is supporting this worth the complexity? Personally, I don't quite think it is. The gains for this API are IMHO super minimal (I actually struggle to name any technical benefit for this) - from my POV we should stick to a single API to avoid bikeshedding in teams & type complexity.

PS. The comment of mine here doesn't even touch on the fact that the first overload is broken as well because it allows silly mistakes like on this [TS playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKIDcLgPICMBWECYyA3gFDLJgCeADhAFzIDOYUoA5gNzkC+5cqEixEKAApwaAGwD2cACYBlLAugAeACoYsJCAA9IIBSzSYcBImAB8ZSsgAUEc2GbaXyAD7J3ugNoA5LQMAQC6AJTM6LLACrwC5MEoOuCmALx2VFRJzAEAYtjYAfZUcCAsAO7QzCAArgC2uND8XplZOcgBAEIAggBKxVnIsmAAFtWs7Fz8gmoI0nBQKAiy5SQsqsySMvLKqhopYCzWggD0p8OydCzkG8YO3f0B4eRAA). It's possible to type this correctly but it would add even more to the overall type complexity.

Reply:
> Is supporting this worth the complexity? Personally, I don't quite think it is.

Yes, it is. There are many developers using the two-argument variant, and I want to continue to support that.

Reply:
For reference, here's the strongly-typed version: [TS Playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKIDcLgPICMBWECYyA3gFDLJgCeADhAFzIDOYUoA5gNzkC+5crQZoAHojAAeAIIAaZACF5AYQB8yALzJpyCKMggAJi0XIA-Kb0Hj288mXJmICJiiPkz171CRYiFAAKcDQANgD2cIYAyliG0JIAKhhYJFaxJsk4BERg6hRUABQuKcxJmOAAlMzoYcCGvFSSANK6+unIZSkA2gDkwhA9ALqqReVgzC0AZGISkqTU9EzILXzyqPpQs53g8vP9E8h8qvJNqlXINXUNyM2t1ibbYL39QyPF4BPydMHhkczYAFtgFJ1uwtpkwLsFgwDkd5H1Fj11OdLvV+IJ+mgxiYtPkqNClj0AGLYbA9Sj4uAgFgAd2gTgArgDcNB+MgAD5kClUfbIHoKaQAJXJ+OQYTAAAt6ax2Fw2Zy8bz+dIAFrkgTkOIIEJwKAoBBhakkFixZhBUIRaKxeIQliqQQmowFZXCirIAD07uQeoAjgzgHqTN8LZEAITkR2GZ0CtUVCOxaNCnpQ8VSqDMHqRHqHOORxPCqFU2nSgAsACYc-GnS7k2QCRmBQWxZLpZnDNm+HGgA)

Reply:
I've played a little bit with implementations from [the article](https://artsy.github.io/blog/2018/11/21/conditional-types-in-typescript/) that we've been talking about and this is what I got:
[TS playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKIDcLgPICMBWECYyA3gFDLJgCeADhAFzIDOYUoA5gNzkC+5crQZoAHojAAeAIIAaZACF5AYQB8yALzJpyCKMggAJi0XIA-Kb0Hj288mXJmICJiiPkz17yH0UqUQgANgCuhhAAKr4y6lqkyADaANLIoGJBoRCSANYQNAD2MNryAETCEMWqALrM0kmVyAI+Iv7sEi1QcAAKcB0AthCQUCwy8uExtlZYNnFlzOENlHb+6WGRDNGLTi7QgmVircQAysC9dIGZ0uM6k0YmcCA0ixakfLr6UybLIatRl4tUFmk-3cnh2VC2rkEoEG8CQyG6NECeTghkOU2gknCGCwJBuNmxOAIRDA6goVAAFNtwHMCWAAJTMdB5YCGXhUAD07OQkmSeJMWMw4HiAHIysLKqpKYKwMxkgAyMQSSQzXyyhrydpKgU4+QqhhqviqeSJVQM5BMlls5BSnHMTVHE5nTLa8CqeKlXzFSpmi2sxac7m8963ZAusAisUSm3U5CJeR0OCI5GGZjYXrAKT2qRh3XUVWxhpG5Ci3zC9Q+5l+qg8t7Wfm0iOlqNUmWx+OJpEou36DrETXdPoDaDDHOx8uMyu8Rp7WkmWLA2bFgBi2GwwuB9xYAHdoE5gr1cDtXgAfMgL-PChTSABK66oVDyYAAFrvWOwuPxkKeyXn9cWrwAWuujRhEEPQoAgeQgGwrBTMwCKdqi6JQJIs6qIILBTOSl7SEBuZwHAzDCgAjMKDR0hhWE4bedLIAGUAQAAjsEwAMSYCZJiiACE5CYUY2GAcKFF8YYAk3sKuaPi+UBESiZF8MJVFXre+HQTuMnIAALAATORvFKeJuaLtREnIFJr7CnJelAA)

In addition - there is this extra requirement~ that for an event without extra params the second param should not be allowed at all. By excluding "simple" events we can easily get it:
[TS playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKIDcLgPICMBWECYyA3gFDLJgCeADhAFzIDOYUoA5gNzkC+5crQZoAHojAAeAIIAaZACF5AYQB8yALzJpyCKMggAJi0XIA-Kb0Hj288mXJmICJiiPkz17yH0UqUQgANgCuhhAAKr4y6lqkyADaANLIoGJBoRCSANYQNAD2MNryAETCEMWqALrM0kmVyAI+Iv7sEi1QcAAKcB0AthCQUCwy8uExtlZYNnFlzOENlHb+6WGRDNGLTi7QgmVircQAysC9dIGZ0uM6k0YmcCA0ixakfLr6UybLIatRl4tUFmk-3cnh2VC2rkEoEG8CQyG6NECeTghkOU2gknCGCwJBuNmxOAIRDA6goVAAFNtwHMCWAAJTMdB5YCGXhUAD07OQkmSeJMWMw4HiAHIysLKqpKYKwMxkgAyMQSSQzXyyhrydpKgU4+QqhhqviqeSJVQM5BMlls5BSnHMTVHE5nTLa8CqeKlXzFSpmi2sxac7m8963ZAusAisUSm3U5CJeR0OCI5GGZjYXrAKT2qRh3XUVWxhpG5Ci3zC9Q+5l+qg8t7WT4Bb7O2kR0uVDX6DoO07nTG0t0ehhe1SSqky2PxxNIlF2jttWfdPoDaDDHOx8uMyu8Rp7WkmWLA2bFgBi2GwwuB9xYAHdoE5gr1cDtXgAfMgH-PChTSABK56oVDyMAAAtb1YdguH4ZBXzJPN9WLL8AC1z0aMIgh6FAEDyEA2FYKZmARKdUXRKBJF3VRBBYKZyU-aQkNzOA4GYYUAEZhQaOkKKomjfzpZAAygCAAEdgmAASTATJMUQAQnISijGoxDhQ4uTDAUn9hVzQCQKgJiUTYvhlK4r9f3o7Cbx05AABYACZ2Nkoz1NzQ9uI05AtNA4U9LsoA)
but I have also found a better way to do this - which also has a better IDE displays when you make an error:
[TS playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKIDcLgPICMBWECYyA3gFDLJgCeADhAFzIDOYUoA5gNzkC+5crQZoAHojAAeAIIAaZACF5AYQB8yALzJpyCKMggAJi0XIA-Kb0Hj288mXJmICJiiPkz17yH0UqUQgANgCuhhAAKr4y6lqkyADaANLIoGJBoRCSANYQNAD2MNryAETCEMWqALrM0kmVyAI+Iv7sEi1QcAAKcB0AthCQUCwy8uExtlZYNnFlzOENlHb+6WGRDNGLTi7QgmVircQAysC9dIGZ0uM6k0YmcCA0ixakfLr6UybLIatRl4tUFmk-3cnh2VC2rl2vmQADltlAAJIwVCnWiSMaaMivG42eYWUFuOaCUCDeBIZDdGiBPJwQyHKbQdEYLAkHGfTA4AhEMDqChUAAU23Ac2Z4AAlMx0HlgIZeFQAPTy5CSZJs5DhUVgeIAcjK2sqqkFHLAzGSADIxBJJDNfKaGvJ2laNcb5DaGHa+Kp5IlVBLkFKZXLkEaWcxHUcTmdMs6Wap4qVfMVKn6A7LForlar3rd1ZqdXqDSHhchEvI6HAqTTDMxsL1gFJw1IY+BXdRbSWGl7kLrfNr1CnpWmqCq3tYTM2tT2GPrDUKTSWyxXqbTmHDXEiUXQ0Y3Hd0+gNoMMJ97VP3JYPeI09pqTLFgbNuwAxbDYbXA+4sADu0CcwV6uB2V4AB8yHvdttQUaQACU3yoKg8jAAALH9WHYLh+GQEC+Tbd1u0ggAtN9GjCIIehQBA8hANhWCmZhKWXOkGSgSQb1UQQWCmfkIOkQjWzgOBmG1ABGbUGjFdjOO4mCxWQDMoAgABHYJgHkkxy0rWkAEJyA4owuII7VxN0wx9Og7VWwQ5CoEE2lRL4IzJMgmC+Ko79rOQAAWAAmMSdMcszWwfKTzOQSyUO1WzfKAA)

Reply:
@Andarist Updated the types: 62e6b90